cumulus/parachains/runtimes/assets/test-utils/src/test_cases_over_bridge.rs

```diff
@@ -0,0 +1,621 @@
+		+ Into<AccountId>,
+	<<Runtime as frame_system::Config>::Lookup as StaticLookup>::Source:
+		From<<Runtime as frame_system::Config>::AccountId>,
+	ForeignAssetsPalletInstance: 'static,
+{
+	ExtBuilder::<Runtime>::default()
+		.with_collators(collator_session_keys.collators())
+		.with_session_keys(collator_session_keys.session_keys())
+		.with_tracing()
+		.build()
+		.execute_with(|| {
+			// Set account as block author, who will receive fees
+			RuntimeHelper::<Runtime, AllPalletsWithoutSystem>::run_to_block(
+				2,
+				block_author_account.clone().into(),
+			);
+
+			// prepare bridge config
+			let TestBridgingConfig { local_bridge_hub_location, .. } = prepare_configuration();
+
+			// drip 'ED' user target account
+			let _ = <pallet_balances::Pallet<Runtime>>::deposit_creating(
+				&target_account,
+				existential_deposit,
+			);
+
+			// sovereign account as foreign asset owner (can be whoever for this scenario, doesnt
+			// matter)
+			let sovereign_account_as_owner_of_foreign_asset =
+				LocationToAccountId::convert_location(&MultiLocation::parent()).unwrap();
+
+			// staking pot account for collecting local native fees from `BuyExecution`
+			let staking_pot = <pallet_collator_selection::Pallet<Runtime>>::account_id();
+			let _ = <pallet_balances::Pallet<Runtime>>::deposit_creating(
+				&staking_pot,
+				existential_deposit,
+			);
+
+			// create foreign asset for wrapped/derivated representation
+			assert_ok!(
+				<pallet_assets::Pallet<Runtime, ForeignAssetsPalletInstance>>::force_create(
+					RuntimeHelper::<Runtime, AllPalletsWithoutSystem>::root_origin(),
+					foreign_asset_id_multilocation.into(),
+					sovereign_account_as_owner_of_foreign_asset.clone().into(),
+					true, // is_sufficient=true
+					foreign_asset_id_minimum_balance.into()
+				)
+			);
+
+			// Balances before
+			assert_eq!(
+				<pallet_balances::Pallet<Runtime>>::free_balance(&target_account),
+				existential_deposit.clone()
+			);
+			assert_eq!(
+				<pallet_balances::Pallet<Runtime>>::free_balance(&block_author_account),
+				0.into()
+			);
+			assert_eq!(
+				<pallet_balances::Pallet<Runtime>>::free_balance(&staking_pot),
+				existential_deposit.clone()
+			);
+
+			// ForeignAssets balances before
+			assert_eq!(
+				<pallet_assets::Pallet<Runtime, ForeignAssetsPalletInstance>>::balance(
+					foreign_asset_id_multilocation.into(),
+					&target_account
+				),
+				0.into()
+			);
+			assert_eq!(
+				<pallet_assets::Pallet<Runtime, ForeignAssetsPalletInstance>>::balance(
+					foreign_asset_id_multilocation.into(),
+					&block_author_account
+				),
+				0.into()
+			);
+			assert_eq!(
+				<pallet_assets::Pallet<Runtime, ForeignAssetsPalletInstance>>::balance(
+					foreign_asset_id_multilocation.into(),
+					&staking_pot
+				),
+				0.into()
+			);
+
+			// Call received XCM execution
+			let xcm = Xcm(vec![
+				DescendOrigin(bridge_instance),
+				UniversalOrigin(universal_origin),
+				DescendOrigin(descend_origin),
```

Comment:
What's up with this? The `UniversalOrigin` instruction will override the previous `DescendOrigin` instruction

Reply:
Yes, exactly, the message comes from BridgeHub through XcmpQueue, so the origin is like `MultiLocation(1, X1(Parachain(1002)))` and the first `DescendOrigin(bridge_instance)` adds it to the origin like:
```
MultiLocation(1, X2(Parachain(1002), bridge_instance))
```
And we also define `MultiLocation(1, X2(Parachain(1002), bridge_instance))`  to the `UniversalAliases`,
which is used for processing `UniversalOrigin` which checks actual origin with `Config::UniversalAliases::contains(`.
(This fix was requested by SA)

```
                   UniversalOrigin(new_global) => {
				let universal_location = Config::UniversalLocation::get();
				ensure!(universal_location.first() != Some(&new_global), XcmError::InvalidLocation);
				let origin = *self.origin_ref().ok_or(XcmError::BadOrigin)?;
				let origin_xform = (origin, new_global);
				let ok = Config::UniversalAliases::contains(&origin_xform);
```